proof-systems/src/darlin/poly-commit/src/domain_extended/constraints/mod.rs

```
@@ -0,0 +1,460 @@
+    PC: PolynomialCommitment<G>,
+    PCG: PolynomialCommitmentVerifierGadget<ConstraintF, G, PC>,
+> {
+    _field: PhantomData<ConstraintF>,
+    _curve: PhantomData<G>,
+    _pc: PhantomData<PC>,
+    _pcg: PhantomData<PCG>,
+}
+
+impl<
+        ConstraintF: PrimeField,
+        G: Group<BaseField = ConstraintF>,
+        PC: 'static + PolynomialCommitment<G, Commitment = G>,
+        PCG: 'static + PolynomialCommitmentVerifierGadget<ConstraintF, G, PC>,
+    > DomainExtendedPolyCommitVerifierGadget<ConstraintF, G, PC, PCG>
+{
+    // this function allows to convert the set of commitments of a segmented polynomial to a single commitment,
+    // which can be used to verify an opening proof for the polynomial at point `point`.
+    // For efficiency, the function allows to simultaneously convert set of commitments
+    // for multiple segmented polynomials with an opening proof for the same point `point`.
+    fn combine_commitments<CS: ConstraintSystemAbstract<ConstraintF>>(
+        mut cs: CS,
+        vk: &PCG::VerifierKey,
+        commitments: &[LabeledCommitmentGadget<
+            Self,
+            ConstraintF,
+            G,
+            DomainExtendedPolynomialCommitment<G, PC>,
+        >],
+        point: &NonNativeFieldGadget<G::ScalarField, ConstraintF>,
+    ) -> Result<Vec<LabeledCommitmentGadget<PCG, ConstraintF, G, PC>>, SynthesisError> {
+        let s = vk.segment_size() + 1;
+        let point_to_s = point.pow_by_constant(cs.ns(|| "point^s"), [s as u64])?;
+        let point_to_s_bits = point_to_s.to_bits_for_normal_form(cs.ns(|| "point^s to bits"))?;
+
+        let mut iterators = Vec::new();
+        let mut labels = Vec::new();
+        for el in commitments {
+            let cloned_el = el.clone();
+            labels.push(cloned_el.label().clone());
+            iterators.push(el.commitment().iter().rev());
+        }
+
+        let zero_commitment = PCG::Commitment::zero(cs.ns(|| "alloc zero commitment"))?;
```

No reason to worry about not having commitments to combine, it would be a misuse. Let's throw an error if this happens

---

Ok